### PR TITLE
Implement promptflow model flavor

### PR DESCRIPTION
**Pull request by brynn-code**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Implement promptflow model flavor

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/promptflow/__init__.py
Comment (wangchao1230) [inline mlflow/promptflow/__init__.py:66]: remvoe this if not used

File: mlflow/promptflow/__init__.py
Comment (wangchao1230) [inline mlflow/promptflow/__init__.py:142]: LLMChain?

File: mlflow/promptflow/__init__.py
Comment (wangchao1230) [inline mlflow/promptflow/__init__.py:219]: update

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:219]: reminder for AK: understand additional_includes vs code_paths behavior

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:321]: below you are passing in a path, but here you are treating it like a rehydrated object

File: tests/promptflow/test_promptflow_model_export.py
Comment (akshaya-a) [inline tests/promptflow/test_promptflow_model_export.py:41]: are there assertions missing here if the test passes with my other comment on the implementation? (path vs object)

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:330]: can you add examples of inputs here

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:339]: if this is needed on every request should you move this initialization of the invoker to the init/constructor

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:356]: any minibatching api we can implement on the invoker side to push down the batching problem?

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:341]: @santiagxf any chance we can put this in inference config?

Comment (akshaya-a) [reply mlflow/promptflow/__init__.py:341]: ideally we can move this stuff into params + inference config

Comment (akshaya-a) [reply mlflow/promptflow/__init__.py:341]: (though, TBH gateway vs connection provider here is probably going to get interesting)

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [inline mlflow/promptflow/__init__.py:225]: minor point: if you instead put a .to_mlflow_signature() in the pf model object directly, iterating on this code would be easier, but that's not a big deal

File: mlflow/promptflow/__init__.py
Comment (wxpjimmy) [inline mlflow/promptflow/__init__.py:232]: is this the suggested way to fall back other types to string?

File: mlflow/promptflow/__init__.py
Comment (wxpjimmy) [reply mlflow/promptflow/__init__.py:341]: we should allow customizing this connection provider here

File: mlflow/promptflow/__init__.py
Comment (wxpjimmy) [reply mlflow/promptflow/__init__.py:356]: @brynn-code I'm assuming we don't need to support batch input here, this is mainly for serving scenario and we can allow only one input message. Or is there any other scenario we need to cover here?

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:225]: Thanks, I prefer leaving them here, because promptflow package doesn't include mlflow in dependencies for now.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:232]: There is no suitable type for list, object, image, etc. At the same time, I saw langchain dump all their inputs/outputs to string type.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:397]: Thanks, refined.

File: tests/promptflow/test_promptflow_model_export.py
Comment (brynn-code) [reply tests/promptflow/test_promptflow_model_export.py:53]: It doesn't fail because FlowInvoker can accept both str and object type, I've refine the parameter name in `__init__.py`.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Good catch, moved.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Added model_config to allow user set connection provider.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: We limited the data size to 1 now if passing data frame, as for our scenario seems user shall request one by one... + @wxpjimmy

File: mlflow/promptflow/__init__.py
Comment (akshaya-a) [reply mlflow/promptflow/__init__.py:306]: hey @brynn-code you might want to sync up with @serena-ruan  on her current work enabling arbitrary input types here: https://github.com/mlflow/mlflow/tree/llm_signature

this is probably the best alignment of your work, but please dig into the details there - at the end of the day the goal of serena's work is to enable stuff like what you're doing, so maybe we can check to see how aligned the design is

File: mlflow/promptflow/__init__.py
Comment (santiagxf) [reply mlflow/promptflow/__init__.py]: @wxpjimmy @brynn-code I saw we changed the implementation to now take the entire request as a single conversation with multiple messages. I think this should work, but let's make it explicit because other flavors in MLflow assume each row is an independent sample (in your case each row is part of the same sample (conversation))

File: mlflow/promptflow/__init__.py
Comment (wxpjimmy) [reply mlflow/promptflow/__init__.py:306]: @akshaya-a seems the arbitrary input type support has not finished yet, I would suggest we handle the signature in two steps:
1. continue this promptflow flavor contributing effort without signature in MLmodel
2. once that arbitrary type supporting is ready, we can update promptflow flavor to add back the signature.

We'd better not introduce hard dependencies between these two efforts.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Yes, we'd like to add this limitation, is there somewhere we need to emphasize it to user?

File: mlflow/promptflow/__init__.py
Comment (serena-ruan) [reply mlflow/promptflow/__init__.py]: llm_signature is planned to release with 2.10.0, and ideally you should be able to reuse whatever we supports within infer_signature directly :)

File: mlflow/ml_package_versions.py
Comment (harupy) [inline mlflow/ml_package_versions.py:258]: any breaking changes between 1.1 and 2.0?

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (harupy) [inline tests/promptflow/flow_with_additional_includes/hello.py:31]:
```suggestion
    logit_bias: dict = None,
```

We should not use a mutable object as a default value.

File: tests/promptflow/flow_with_additional_includes/flow.dag.yaml
Comment (harupy) [inline tests/promptflow/flow_with_additional_includes/flow.dag.yaml:31]:
```suggestion
  - ../additional_file

```

File: tests/promptflow/flow_with_additional_includes/hello.jinja2
Comment (harupy) [inline tests/promptflow/flow_with_additional_includes/hello.jinja2:2]:
```suggestion
Write a simple {{text}} program that displays the greeting message when executed.

```

File: mlflow/ml_package_versions.py
Comment (harupy) [reply mlflow/ml_package_versions.py:258]: looks like 2.0 is not released yet: https://pypi.org/project/promptflow/#history

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (harupy) [inline tests/promptflow/flow_with_additional_includes/hello.py:29]: Do we really need these arguments?

Comment (harupy) [reply tests/promptflow/flow_with_additional_includes/hello.py:29]: Can we reduce them to a few?

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (wangchao1230) [reply tests/promptflow/flow_with_additional_includes/hello.py:42]: We should update this to use chat api

File: mlflow/ml_package_versions.py
Comment (brynn-code) [reply mlflow/ml_package_versions.py:258]: is it ok if we don't add the maximum? seems all the package have both minimum and maximum

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (brynn-code) [reply tests/promptflow/flow_with_additional_includes/hello.py:37]: I've removed that, btw, what's the openai package version are we using in CI? I met import error when trying to run tests locally, it seems 'openai.error' is used somewhere.

Comment (brynn-code) [reply tests/promptflow/flow_with_additional_includes/hello.py:37]: Oh, I get it, I'll update this to a mock response

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (brynn-code) [reply tests/promptflow/flow_with_additional_includes/hello.py]: Updated to a simple python flow.

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (brynn-code) [reply tests/promptflow/flow_with_additional_includes/hello.py]: Tested and removed.

File: tests/promptflow/flow_with_additional_includes/hello.py
Comment (brynn-code) [reply tests/promptflow/flow_with_additional_includes/hello.py]: Updated to use openai 1.x client.

File: mlflow/ml_package_versions.py
Comment (brynn-code) [reply mlflow/ml_package_versions.py]: updated.

File: tests/promptflow/test_promptflow_model_export.py
Comment (serena-ruan) [inline tests/promptflow/test_promptflow_model_export.py:14]: Do we have to keep the folder in order to test? Is there any functions that you could directly save a flow like this?

File: tests/promptflow/test_promptflow_model_export.py
Comment (serena-ruan) [inline tests/promptflow/test_promptflow_model_export.py:31]: Could we test on loaded_model.predict(...) and compare to the original model's prediction result

File: tests/promptflow/test_promptflow_model_export.py
Comment (serena-ruan) [inline tests/promptflow/test_promptflow_model_export.py:66]: Let's also add spark_udf & model serving tests. You could check other flavor's tests as reference.

File: tests/promptflow/test_promptflow_model_export.py
Comment (serena-ruan) [inline tests/promptflow/test_promptflow_model_export.py:46]: We could combine this test with test_promptflow_model_predict

File: tests/promptflow/test_promptflow_model_export.py
Comment (serena-ruan) [reply tests/promptflow/test_promptflow_model_export.py:66]: Also could we add more examples to make sure they work as well? I'm not sure if there're different model types promptflow supports, but let's make sure test coverage is enough.

File: tests/promptflow/test_promptflow_model_export.py
Comment (brynn-code) [reply tests/promptflow/test_promptflow_model_export.py:14]: Yes, we can't write a flow directly using python from scratch for now, need to keep this folder.

File: tests/promptflow/test_promptflow_model_export.py
Comment (brynn-code) [reply tests/promptflow/test_promptflow_model_export.py:41]: added

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Offline synced and add _infer_signature_from_input_example.

File: tests/promptflow/test_promptflow_model_export.py
Comment (brynn-code) [reply tests/promptflow/test_promptflow_model_export.py]: updated

File: tests/promptflow/test_promptflow_model_export.py
Comment (brynn-code) [reply tests/promptflow/test_promptflow_model_export.py:95]: Thanks, test added. Promptflow only have the Flow entity to be wrapped as a MLflow model, so no any other types.

File: mlflow/promptflow/__init__.py
Comment (B-Step62) [inline mlflow/promptflow/__init__.py:461]: Do we really need`predict` API? `load_model()` API is intended to load the underlying model, if users want to have `predict` API then they should use pyfunc load. Also I don't think we want users to write production code relying on this patch:)

File: tests/promptflow/test_promptflow_model_export.py
Comment (B-Step62) [inline tests/promptflow/test_promptflow_model_export.py:60]: Let's split this to multiple test cases. Bundling all cases into one prevents quick iteration, e.g. none of the checks until the last assertion need spark session.

**Review by B-Step62 — COMMENTED** [review]
Thank you so much for this massive contribution! Left a few minor comments, but generally looks good.

Also it seems the linter job failed, could you run `pre-commit run --all-files` before pushing the code to make sure everything follows the style standard. (btw python/window jobs failure is not relevant to this change; we are addressing the issue so please ignore them.)

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:378]: Thanks, I'd like to keep the original one, because default argument value shall not be mutable, and {} is a mutable one.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:277]: Feels like it would be clearer when path exists, user will see `Path 'C:/users/xx..../target'` already exists, rather than 'target' already exists, which is a little vague.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:78]: Yes, promptflow-tools includes built-in tools provided by promptflow, which is not a mandatory package, if user has a flow with all nodes write by themselves, tools is not required for executing the flow.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Got it, removed.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py:57]: removed.

File: mlflow/promptflow/__init__.py
Comment (B-Step62) [reply mlflow/promptflow/__init__.py:279]: Makes a lot sense! We may add this to the utility itself later, thanks.

File: mlflow/promptflow/__init__.py
Comment (brynn-code) [reply mlflow/promptflow/__init__.py]: Thank you. Comments added, and we could say 'yes' for the statement, if in dev environment running the flow not require promptflow-tools package, then serving/deployment doesn't need it either.

**Review by B-Step62 — APPROVED** [review]
LGTM! Thank you for reflecting all comments and discussion!

**Comment by brynn-code** [comment]
I don't have write access the the repository, please help to merge it if someone see this, thanks in advance!

**Comment by B-Step62** [comment]
@brynn-code May I ask one follow-up for this PR, could you add a section for promptflow flavor in the [Build-In Model Flavors](https://mlflow.org/docs/latest/models.html#built-in-model-flavors) doc? It can be just a short description for the flavor and an example code snippet. Also I think we should note the signature limitation. This PR already adds a page for `mlflow.promptflow` but adding to the list would increase visibility of the flavor a lot. Thanks!

**Comment by B-Step62** [comment]
@brynn-code, Unfortunately, we needed to revert this PR to unblock the release. This PR didn't pass CI because the tests in `test_promptflow_model_export` failed. 

The main issue is that `promptflow-tools` requires OpenAI >= 1.0, which MLflow doesn't support now. It turns out the tools is not optional for making prediction, because FlowInvoker indirectly uses the tool [here](https://github.com/microsoft/promptflow/blob/main/src/promptflow/promptflow/executor/_tool_resolver.py#L249). We plan to work on OpenAI support early in a few weeks. We will let you know once it's done and then revive the flavor.

Aside from this, there are a few other minor issues in unit tests:
  - A fixture `spark` is used as a constant [here](https://github.com/mlflow/mlflow/blob/master/tests/promptflow/test_promptflow_model_export.py#L105), which caused the test failure. The correct usage is to pass as a param of test method like `def test_promptflow_model_sparkudf_predict(spark):`
  - Regex pattern [here](https://github.com/mlflow/mlflow/blob/master/tests/promptflow/test_promptflow_model_export.py#L119) to [the actual message](https://github.com/mlflow/mlflow/blob/master/mlflow/promptflow/__init__.py#L58-L59).

Sadly these issues were not captured in the original PR because the github action didn't run those flavor tests, due to some misconfiguration - namely, promptflow was added to `ml_package_version.py` instead of `ml-package-versions.yml` (the latter is what github action looks at, while the former is auto-generated file based on it). This is not your fault as the configuration is complicated and we also overlooked it, but I'd kindly request running unit tests locally for double-check ([test guide](https://github.com/B-Step62/mlflow/blob/master/CONTRIBUTING.md#python)). Thanks.

**Comment by brynn-code** [comment]
@B-Step62 The FlowInvoker change caused by our recently separating packages activities, we will discussion if we need to make 1.7.0 compatible with the old version, or update promptflow flavor code and promptflow version in ml-package-versions.yml to fit those changes after promptflow 1.7.0 release (we scheduled to release it today), I'll reply here as soon.

And for the new changes afterwards, would you mind open issues here https://github.com/microsoft/promptflow/issues so that we won't miss it? We will triage the github issue daily, I'm afraid I'll miss the PR notification, issue may be more visible, thank you!

**Comment by brynn-code** [comment]
@B-Step62 We have fix the issue, to ensure backward compatibility, thanks for your time, really appreciate it.

**Comment by B-Step62** [comment]
We confirmed the latest commit resolves the issue, thank you so much for the quick fix @brynn-code! 

> And for the new changes afterwards, would you mind open issues here https://github.com/microsoft/promptflow/issues so that we won't miss it?

Absolutely! I will contact by filing an issue in the promptflow repo next time.

**Pull request merged**
